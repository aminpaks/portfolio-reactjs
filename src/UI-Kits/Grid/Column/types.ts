import { CSSProperties } from 'react';

export type ColumnSize = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | null;
export interface ColumnBreakpointSize {
  xs?: ColumnSize;
  sm?: ColumnSize;
  md?: ColumnSize;
  lg?: ColumnSize;
  xl?: ColumnSize;
  xxl?: ColumnSize;
  size?: ColumnSize;
}
export type BreakpointGroupSize = { [breakpoint: string]: Array<number | null> };
export interface GridColumnProps extends ColumnBreakpointSize {
  style?: CSSProperties;
}
export interface InternalGridColumnProps {
  size: ColumnBreakpointSize;
  allColumnSizes: BreakpointGroupSize;
  style?: CSSProperties;
}
