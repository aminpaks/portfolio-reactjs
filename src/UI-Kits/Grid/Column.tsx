import React, { FC, CSSProperties } from 'react';
import { StyledGridColumnContainer } from './Column.styled';

export interface GridColumnProps {
  size: number | null;
  allSizes: (number | null)[];
  style?: CSSProperties;
}
export type PGridColumnProps = Partial<GridColumnProps>;

export const Column: FC<PGridColumnProps> = ({ children, size = null, allSizes = [], style }) => (
  <StyledGridColumnContainer size={size} allSizes={allSizes} style={style}>
    {children}
  </StyledGridColumnContainer>
);
