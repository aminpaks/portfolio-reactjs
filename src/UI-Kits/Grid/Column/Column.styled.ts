import styled from 'src/Theme';

import { InternalGridColumnProps } from './types';
import { getSizeValue } from './Util';

export const StyledGridColumnContainer = styled.div<InternalGridColumnProps>`
  display: block;
  box-sizing: border-box;
  flex: 0 0 auto;
  flex-basis: ${getSizeValue('sm')};
  max-width: ${getSizeValue('sm')};

  @media (min-width: ${props => props.theme.breakpoints.sm}px) {
    flex-basis: ${getSizeValue('md')};
    max-width: ${getSizeValue('md')};
  }
  @media (min-width: ${props => props.theme.breakpoints.md}px) {
    flex-basis: ${getSizeValue('lg')};
    max-width: ${getSizeValue('lg')};
  }
  @media (min-width: ${props => props.theme.breakpoints.lg}px) {
    flex-basis: ${getSizeValue('xl')};
    max-width: ${getSizeValue('xl')};
  }
  @media (min-width: ${props => props.theme.breakpoints.xl}px) {
    flex-basis: ${getSizeValue('xxl')};
    max-width: ${getSizeValue('xxl')};
  }

  &:last-of-type {
    margin-left: auto;
  }
`;
