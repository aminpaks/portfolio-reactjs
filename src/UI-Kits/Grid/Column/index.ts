export * from './Column';
export * from './Column.styled';
export * from './types';
