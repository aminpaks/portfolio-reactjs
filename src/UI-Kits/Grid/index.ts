export * from './Column';
export * from './Grid';
export * from './Grid.styled';
