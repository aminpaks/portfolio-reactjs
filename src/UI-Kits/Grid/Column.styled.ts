import styled from 'src/Theme';
import { GridColumnProps } from './Column';

const getPotentialSize = (sizes: (number | null)[]) => {
  const totalSize = sizes.reduce<number>((res, s) => res + (s || 0), 0);
  const total = sizes.filter(v => v == null);
  const x = (12 - totalSize) / total.length;
  return x;
};

const getSizeValue = ({ size: potentialSize, allSizes: sizes }: GridColumnProps) => {
  const size = potentialSize == null ? getPotentialSize(sizes) : potentialSize;
  const percentage = (100 / 12) * size;
  return percentage <= 0 ? 'auto' : `${percentage}%`;
};

export const StyledGridColumnContainer = styled.div<GridColumnProps>`
  display: block;
  box-sizing: border-box;
  flex: 0 0 auto;
  flex-basis: ${getSizeValue};
  max-width: ${getSizeValue};

  &:not(:first-of-type) {
    margin-left: auto;
  }
`;
