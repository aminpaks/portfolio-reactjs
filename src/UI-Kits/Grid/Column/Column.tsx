import React, { FC } from 'react';
import { StyledGridColumnContainer } from './Column.styled';
import { InternalGridColumnProps } from './types';

export const Column: FC<InternalGridColumnProps> = ({ size, allColumnSizes, style, children }) => (
  <StyledGridColumnContainer size={size} allColumnSizes={allColumnSizes} style={style} data-testid="uikits-grid-column">
    {children}
  </StyledGridColumnContainer>
);
