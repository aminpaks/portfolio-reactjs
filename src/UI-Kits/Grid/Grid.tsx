import React, { FC, ReactElement, Children } from 'react';

import { isChildTypeOf } from '../Utils';
import { Column, PGridColumnProps } from './Column';
import { StyledGridContainer } from './Grid.styled';

export interface GridProps {
  gutter: number;
  unit: string;
  children: ReactElement<PGridColumnProps> | Array<ReactElement<PGridColumnProps> | null> | null;
}

export type PGridProps = Partial<GridProps>;

export const Grid: FC<PGridProps> & { Col: typeof Column } = ({ gutter = 2, unit = 'em', children }) => {
  // Calculate all the children's sizes in an array
  const allSizes: (number | null)[] = [];
  Children.forEach(children, child => {
    if (isChildTypeOf(Column, child)) {
      allSizes.push(child.props.size || null);
    } else {
      allSizes.push(null);
    }
  });
  // Clone all the children with total sizes as prop
  const updatedChildren = Children.map(children, child =>
    isChildTypeOf(Column, child)
      ? React.cloneElement(child, {
          allSizes,
        })
      : null,
  );

  return (
    <StyledGridContainer gutter={gutter} unit={unit}>
      {updatedChildren}
    </StyledGridContainer>
  );
};
// Static property
Grid.Col = Column;
