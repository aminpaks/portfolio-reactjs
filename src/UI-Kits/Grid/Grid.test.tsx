import 'jest-styled-components';

import React from 'react';
import TestRenderer, { ReactTestRenderer } from 'react-test-renderer';
import { ThemeProvider } from 'src/Theme';

import { Column } from './Column';
import { Grid } from './Grid';

describe('Grid', () => {
  let renderer: ReactTestRenderer;

  afterEach(() => {
    if (renderer) {
      renderer.unmount();
    }
  });

  it.only('should render Grid with Columns', () => {
    renderer = TestRenderer.create(
      <ThemeProvider>
        <Grid>
          <Grid.Col sm={6} md={3}>
            column1
          </Grid.Col>
          {/* DIV & SPAN will get ignored */}
          <div />
          <span>column1.2</span>
          <Grid.Col sm={6} md={3}>
            column2
          </Grid.Col>
          <Grid.Col sm={6} md={3}>
            column3
          </Grid.Col>
          <Grid.Col sm={6} md={3}>
            column4
          </Grid.Col>
          {/* DIV will get ignored */}
          <div />
        </Grid>
      </ThemeProvider>,
    );

    const instance = renderer.root;
    const allColumnChildren = instance.findAll(node => node.type === Column);

    expect(allColumnChildren.length).toBe(4);
    expect(allColumnChildren[0].props.children).toBe('column1');
    expect(allColumnChildren[0].props.size).toEqual({
      xs: null,
      sm: 6,
      md: 3,
      lg: null,
      xl: null,
      xxl: null,
      size: null,
    });
    expect(allColumnChildren[0].props.allColumnSizes).toEqual({
      xs: [null, null, null, null],
      sm: [6, 6, 6, 6],
      md: [3, 3, 3, 3],
      lg: [null, null, null, null],
      xl: [null, null, null, null],
      xxl: [null, null, null, null],
      size: [null, null, null, null],
    });
    expect(allColumnChildren[1].props.children).toBe('column2');
    expect(allColumnChildren[2].props.children).toBe('column3');
    expect(allColumnChildren[3].props.children).toBe('column4');

    expect(renderer.toJSON()).toMatchSnapshot();
  });
});
