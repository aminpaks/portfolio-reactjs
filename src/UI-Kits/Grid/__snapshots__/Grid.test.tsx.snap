// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid should render Grid with Columns 1`] = `
.c2 {
  display: block;
  box-sizing: border-box;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  -webkit-flex-basis: 50%;
  -ms-flex-preferred-size: 50%;
  flex-basis: 50%;
  max-width: 50%;
}

.c2:last-of-type {
  margin-left: auto;
}

.c3 {
  display: none;
  box-sizing: border-box;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  -webkit-flex-basis: 0;
  -ms-flex-preferred-size: 0;
  flex-basis: 0;
  max-width: 0;
}

.c3:last-of-type {
  margin-left: auto;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 0 1 auto;
  -ms-flex: 0 1 auto;
  flex: 0 1 auto;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  box-sizing: border-box;
  margin-left: -1.14rem;
  margin-right: -1.14rem;
}

.c0 .c1 {
  padding-left: 1.14rem;
  padding-right: 1.14rem;
}

@media (min-width:576px) {
  .c2 {
    display: block;
    -webkit-flex-basis: 25%;
    -ms-flex-preferred-size: 25%;
    flex-basis: 25%;
    max-width: 25%;
  }
}

@media (min-width:768px) {
  .c2 {
    display: block;
    -webkit-flex-basis: 25%;
    -ms-flex-preferred-size: 25%;
    flex-basis: 25%;
    max-width: 25%;
  }
}

@media (min-width:992px) {
  .c2 {
    display: block;
    -webkit-flex-basis: 25%;
    -ms-flex-preferred-size: 25%;
    flex-basis: 25%;
    max-width: 25%;
  }
}

@media (min-width:1200px) {
  .c2 {
    display: block;
    -webkit-flex-basis: 25%;
    -ms-flex-preferred-size: 25%;
    flex-basis: 25%;
    max-width: 25%;
  }
}

@media (min-width:576px) {
  .c3 {
    display: none;
    -webkit-flex-basis: 0;
    -ms-flex-preferred-size: 0;
    flex-basis: 0;
    max-width: 0;
  }
}

@media (min-width:768px) {
  .c3 {
    display: block;
    -webkit-flex-basis: 25%;
    -ms-flex-preferred-size: 25%;
    flex-basis: 25%;
    max-width: 25%;
  }
}

@media (min-width:992px) {
  .c3 {
    display: block;
    -webkit-flex-basis: 25%;
    -ms-flex-preferred-size: 25%;
    flex-basis: 25%;
    max-width: 25%;
  }
}

@media (min-width:1200px) {
  .c3 {
    display: block;
    -webkit-flex-basis: 25%;
    -ms-flex-preferred-size: 25%;
    flex-basis: 25%;
    max-width: 25%;
  }
}

<div
  className="c0"
>
  <div
    className="c1 c2"
    data-testid="uikits-grid-column"
    size={
      Object {
        "lg": null,
        "md": 3,
        "size": null,
        "sm": 6,
        "xl": null,
        "xs": null,
        "xxl": null,
      }
    }
  >
    column1
  </div>
  <div
    className="c1 c2"
    data-testid="uikits-grid-column"
    size={
      Object {
        "lg": null,
        "md": 3,
        "size": null,
        "sm": 6,
        "xl": null,
        "xs": null,
        "xxl": null,
      }
    }
  >
    column2
  </div>
  <div
    className="c1 c3"
    data-testid="uikits-grid-column"
  >
    column3
  </div>
  <div
    className="c1 c3"
    data-testid="uikits-grid-column"
  >
    column4
  </div>
</div>
`;
